<a href="https://colab.research.google.com/github/viniciusacosta/rent_calculator/blob/main/data_cleaning/csv_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **1. Carregando Dados**

### **1.1 Importando bibliotecas necessárias**


In [1]:
# Importando bibliotecas, para a manipulação e exploração dos conjuntos de dados.
import pandas as pd

# Importando bibliotecas, para a plotagem de gráficos interativos com o plotly.

import plotly.offline as py
import plotly.graph_objs as go
import plotly.figure_factory as ff

# Importando bibliotecas, para a plotagem de gráficos com o Seaborn e Matplotlib.
import matplotlib.pyplot as plt
import seaborn as sns

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### **Carregando Dados**

In [3]:
# Carregando conjunto de dados.
df = pd.read_csv('/content/drive/MyDrive/data_science/imoveis/apto_rawdata.csv')

# Exibindo as primeiras linhas do DataFrame.
df

Unnamed: 0                                            address     city  \
0               0  Rua Professor José Freitas de Andrade, 428 - C...  aracaju   
1               1    Rua Napoleão Dórea, 723 - Atalaia, Aracaju - SE  aracaju   
2               2  Rua Maria Vasconcelos de Andrade, 1720 - Aruan...  aracaju   
3               3                                Luzia, Aracaju - SE  aracaju   
4               4  Avenida São João Batista, 638 - Ponto Novo, Ar...  aracaju   
...           ...                                                ...      ...   
17238       17238             Rua 9, 544 - Setor Oeste, Goiânia - GO  goiania   
17239       17239             Rua 9, 544 - Setor Oeste, Goiânia - GO  goiania   
17240       17240             Rua 9, 544 - Setor Oeste, Goiânia - GO  goiania   
17241       17241             Rua 9, 544 - Setor Oeste, Goiânia - GO  goiania   
17242       17242             Rua 9, 544 - Setor Oeste, Goiânia - GO  goiania   

       area      rooms    bathrooms  garage  \
0      99.0  4 Quartos  2 Banheiros  1 Vaga   
1      70.0  3 Quartos  2 Banheiros  1 Vaga   
2      80.0  3 Quartos  2 Banheiros  1 Vaga   
3      67.0  2 Quartos   1 Banheiro  1 Vaga   
4      70.0  3 Quartos   1 Banheiro  1 Vaga   
...     ...        ...          ...     ...   
17238   NaN        NaN          NaN     NaN   
17239   NaN        NaN          NaN     NaN   
17240   NaN        NaN          NaN     NaN   
17241   NaN        NaN          NaN     NaN   
17242   NaN        NaN          NaN     NaN   

                                                amenites       price_rent  \
0      Mobiliado Condomínio_fechado Aceita_animais Ar...    R$ 1.500 /mês   
1      Piscina Mobiliado Elevador Aceita_animais Ar-c...  R$ 220.000 /mês   
2      Piscina Churrasqueira Elevador Condomínio_fech...      R$ 750 /mês   
3      Piscina Churrasqueira Playground Salão_de_fest...    R$ 1.250 /mês   
4      Piscina Mobiliado Churrasqueira Condomínio_fec...    R$ 1.200 /mês   
...                                                  ...              ...   
17238                                               None             None   
17239                                               None             None   
17240                                               None             None   
17241                                               None             None   
17242                                               None             None   

        condo        crawled_at  
0         NaN  2022-06-23 21:10  
1      R$ 450  2022-06-23 21:10  
2      R$ 400  2022-06-23 21:10  
3      R$ 395  2022-06-23 21:10  
4      R$ 380  2022-06-23 21:10  
...       ...               ...  
17238     NaN  2022-06-23 22:21  
17239     NaN  2022-06-23 22:21  
17240     NaN  2022-06-23 22:21  
17241     NaN  2022-06-23 22:21  
17242     NaN  2022-06-23 22:21  

[17243 rows x 11 columns]

In [12]:
# Carregando conjunto de dados.
df = pd.read_csv('https://raw.githubusercontent.com/viniciusacosta/rent_calculator/main/scraping/apto_rawdata.csv')

# Exibindo as primeiras linhas do DataFrame.
df.head()

Unnamed: 0                                            address     city  \
0           0  Rua Professor José Freitas de Andrade, 428 - C...  aracaju   
1           1    Rua Napoleão Dórea, 723 - Atalaia, Aracaju - SE  aracaju   
2           2  Rua Maria Vasconcelos de Andrade, 1720 - Aruan...  aracaju   
3           3                                Luzia, Aracaju - SE  aracaju   
4           4  Avenida São João Batista, 638 - Ponto Novo, Ar...  aracaju   

   area      rooms    bathrooms  garage  \
0  99.0  4 Quartos  2 Banheiros  1 Vaga   
1  70.0  3 Quartos  2 Banheiros  1 Vaga   
2  80.0  3 Quartos  2 Banheiros  1 Vaga   
3  67.0  2 Quartos   1 Banheiro  1 Vaga   
4  70.0  3 Quartos   1 Banheiro  1 Vaga   

                                            amenites       price_rent   condo  \
0  Mobiliado Condomínio_fechado Aceita_animais Ar...    R$ 1.500 /mês     NaN   
1  Piscina Mobiliado Elevador Aceita_animais Ar-c...  R$ 220.000 /mês  R$ 450   
2  Piscina Churrasqueira Elevador Condomínio_fech...      R$ 750 /mês  R$ 400   
3  Piscina Churrasqueira Playground Salão_de_fest...    R$ 1.250 /mês  R$ 395   
4  Piscina Mobiliado Churrasqueira Condomínio_fec...    R$ 1.200 /mês  R$ 380   

         crawled_at  
0  2022-06-23 21:10  
1  2022-06-23 21:10  
2  2022-06-23 21:10  
3  2022-06-23 21:10  
4  2022-06-23 21:10

In [13]:
# Verificando as dimensões do dataset.
df.shape

(17243, 11)

Verificamos a existência de **11** variáveis, e **17243** observações dentro do dataset.

In [14]:
# Verificando o nome das colunas
df.columns

Index(['Unnamed: 0', 'address', 'city', 'area', 'rooms', 'bathrooms', 'garage',
       'amenites', 'price_rent', 'condo', 'crawled_at'],
      dtype='object')

In [17]:
# Eliminando colunas indesejadas
df.drop(columns=['Unnamed: 0', 'amenites'], inplace=True)

In [18]:
# Verificando o número de NaNs existentes dentro do dataset
df.isna().sum()

address           0
city              0
area          10119
rooms         10151
bathrooms     10192
garage        10227
price_rent        0
condo         12501
crawled_at        0
dtype: int64

In [19]:
# Eliminando as linhas que não possuem valores price
df.dropna(subset=['price_rent', 'rooms', 'bathrooms', 'area', 'garage', 'condo'], inplace = True)

# Verificando o número de NaNs restantes
df.isna().sum()

address       0
city          0
area          0
rooms         0
bathrooms     0
garage        0
price_rent    0
condo         0
crawled_at    0
dtype: int64

In [20]:
# Criando uma bool series
bool_series = df[['address', 'rooms', 'bathrooms', 'garage', 'area']].duplicated(keep = False)

# Passando valores que não estão na matriz, para vermos somente os valores únicos
df = df[~bool_series]

# Retornando o dataframe
df

address     city   area  \
2      Rua Maria Vasconcelos de Andrade, 1720 - Aruan...  aracaju   80.0   
3                                    Luzia, Aracaju - SE  aracaju   67.0   
4      Avenida São João Batista, 638 - Ponto Novo, Ar...  aracaju   70.0   
5      Rua Matilde Silva Lima, 421 - Luzia, Aracaju - SE  aracaju   65.0   
6                                    Luzia, Aracaju - SE  aracaju   59.0   
...                                                  ...      ...    ...   
17028                         Jardim Goiás, Goiânia - GO  goiania   36.0   
17064  Avenida Rio Negro - Parque Amazônia, Goiânia - GO  goiania   64.0   
17099     Avenida T 11, 2572 - Setor Bueno, Goiânia - GO  goiania   90.0   
17137                          Setor Oeste, Goiânia - GO  goiania  173.0   
17171                        Setor Marista, Goiânia - GO  goiania   43.0   

           rooms    bathrooms   garage     price_rent   condo  \
2      3 Quartos  2 Banheiros   1 Vaga    R$ 750 /mês  R$ 400   
3      2 Quartos   1 Banheiro   1 Vaga  R$ 1.250 /mês  R$ 395   
4      3 Quartos   1 Banheiro   1 Vaga  R$ 1.200 /mês  R$ 380   
5      3 Quartos   1 Banheiro   1 Vaga    R$ 800 /mês  R$ 375   
6      3 Quartos   1 Banheiro   1 Vaga    R$ 850 /mês  R$ 370   
...          ...          ...      ...            ...     ...   
17028   1 Quarto   1 Banheiro   1 Vaga  R$ 2.950 /mês  R$ 320   
17064  2 Quartos  2 Banheiros   1 Vaga  R$ 1.200 /mês  R$ 100   
17099  2 Quartos   1 Banheiro   1 Vaga    R$ 880 /mês  R$ 410   
17137  2 Quartos  3 Banheiros  2 Vagas  R$ 1.300 /mês  R$ 480   
17171   1 Quarto   1 Banheiro   1 Vaga  R$ 2.490 /mês  R$ 340   

             crawled_at  
2      2022-06-23 21:10  
3      2022-06-23 21:10  
4      2022-06-23 21:10  
5      2022-06-23 21:10  
6      2022-06-23 21:10  
...                 ...  
17028  2022-06-23 22:20  
17064  2022-06-23 22:20  
17099  2022-06-23 22:20  
17137  2022-06-23 22:20  
17171  2022-06-23 22:20  

[2659 rows x 9 columns]

In [21]:
#Criando coluna estado
df["estado"] = df["address"].str.split(" - ").str[-1]

df

address     city   area  \
2      Rua Maria Vasconcelos de Andrade, 1720 - Aruan...  aracaju   80.0   
3                                    Luzia, Aracaju - SE  aracaju   67.0   
4      Avenida São João Batista, 638 - Ponto Novo, Ar...  aracaju   70.0   
5      Rua Matilde Silva Lima, 421 - Luzia, Aracaju - SE  aracaju   65.0   
6                                    Luzia, Aracaju - SE  aracaju   59.0   
...                                                  ...      ...    ...   
17028                         Jardim Goiás, Goiânia - GO  goiania   36.0   
17064  Avenida Rio Negro - Parque Amazônia, Goiânia - GO  goiania   64.0   
17099     Avenida T 11, 2572 - Setor Bueno, Goiânia - GO  goiania   90.0   
17137                          Setor Oeste, Goiânia - GO  goiania  173.0   
17171                        Setor Marista, Goiânia - GO  goiania   43.0   

           rooms    bathrooms   garage     price_rent   condo  \
2      3 Quartos  2 Banheiros   1 Vaga    R$ 750 /mês  R$ 400   
3      2 Quartos   1 Banheiro   1 Vaga  R$ 1.250 /mês  R$ 395   
4      3 Quartos   1 Banheiro   1 Vaga  R$ 1.200 /mês  R$ 380   
5      3 Quartos   1 Banheiro   1 Vaga    R$ 800 /mês  R$ 375   
6      3 Quartos   1 Banheiro   1 Vaga    R$ 850 /mês  R$ 370   
...          ...          ...      ...            ...     ...   
17028   1 Quarto   1 Banheiro   1 Vaga  R$ 2.950 /mês  R$ 320   
17064  2 Quartos  2 Banheiros   1 Vaga  R$ 1.200 /mês  R$ 100   
17099  2 Quartos   1 Banheiro   1 Vaga    R$ 880 /mês  R$ 410   
17137  2 Quartos  3 Banheiros  2 Vagas  R$ 1.300 /mês  R$ 480   
17171   1 Quarto   1 Banheiro   1 Vaga  R$ 2.490 /mês  R$ 340   

             crawled_at estado  
2      2022-06-23 21:10     SE  
3      2022-06-23 21:10     SE  
4      2022-06-23 21:10     SE  
5      2022-06-23 21:10     SE  
6      2022-06-23 21:10     SE  
...                 ...    ...  
17028  2022-06-23 22:20     GO  
17064  2022-06-23 22:20     GO  
17099  2022-06-23 22:20     GO  
17137  2022-06-23 22:20     GO  
17171  2022-06-23 22:20     GO  

[2659 rows x 10 columns]

In [22]:
# Contabilizando o número de valores únicos em cada variável do dataset.
info = df.nunique().sort_values()

# Determinando o tipo de dado de cada uma das variáveis do dataset.
info = pd.DataFrame(info.values, index = info.index, columns = ['NUniques'])

# Atribuindo informações sobre o tipo de dado das variáveis ao DataFrame.
info['dtypes'] = df.dtypes

# Exibindo informações.
info

NUniques   dtypes
rooms              7   object
bathrooms          8   object
garage            11   object
city              17   object
estado            17   object
crawled_at        68   object
area             258  float64
price_rent       380   object
condo            776   object
address         2282   object

**Transformando os tipos de dados das colunas**

In [23]:
#### Transformando tipo de dado  das colunas
### Object to float64 - Apartamento e garagem
apto_garage = ["rooms", 'bathrooms', 'garage']

for col in apto_garage:
  df[col] = df[col].str.split(" ").str[0].str.replace("--","0").astype(float)

### Object to float64 - Aluguel, preço de venda e condominio
## Object to float64 - condominio
df["condo"] = df["condo"].str.split(" ").str[1].str.replace(".","").astype(float)

## Object to float64 - Preço aluguel
# Criando função para transformar lista em string
def to_string(x):
  for n in x: 
    string = ''.join(str(n) for n in x)
    return string

# Object to float64 - Preço aluguel
df['price_rent'] = df['price_rent'].str.findall('[0-9]+').apply(to_string).astype(float)

### Object to datetime
df['crawled_at'] = pd.to_datetime(df['crawled_at'])

# Verificando dtypes
info = df.nunique().sort_values()
info = pd.DataFrame(info.values, index = info.index, columns = ['NUniques'])
info['dtypes'] = df.dtypes
info

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



NUniques          dtypes
rooms              7         float64
bathrooms          8         float64
garage            11         float64
city              17          object
estado            17          object
crawled_at        68  datetime64[ns]
area             258         float64
price_rent       380         float64
condo            776         float64
address         2282          object

In [27]:
# Ordenando as colunas
df = df[['crawled_at', 'address', 'estado', 'city', 'area', 'rooms', 'bathrooms', 'garage',
       'condo', 'price_rent']]

In [28]:
# Retirando todos os acentos e deixando todas as letras minusculas para facilitar a manipulação do conjunto de dados
df.columns = df.columns.str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8').str.lower()

In [29]:
df

crawled_at                                            address  \
2     2022-06-23 21:10:00  Rua Maria Vasconcelos de Andrade, 1720 - Aruan...   
3     2022-06-23 21:10:00                                Luzia, Aracaju - SE   
4     2022-06-23 21:10:00  Avenida São João Batista, 638 - Ponto Novo, Ar...   
5     2022-06-23 21:10:00  Rua Matilde Silva Lima, 421 - Luzia, Aracaju - SE   
6     2022-06-23 21:10:00                                Luzia, Aracaju - SE   
...                   ...                                                ...   
17028 2022-06-23 22:20:00                         Jardim Goiás, Goiânia - GO   
17064 2022-06-23 22:20:00  Avenida Rio Negro - Parque Amazônia, Goiânia - GO   
17099 2022-06-23 22:20:00     Avenida T 11, 2572 - Setor Bueno, Goiânia - GO   
17137 2022-06-23 22:20:00                          Setor Oeste, Goiânia - GO   
17171 2022-06-23 22:20:00                        Setor Marista, Goiânia - GO   

      estado     city   area  rooms  bathrooms  garage  condo  price_rent  
2         SE  aracaju   80.0    3.0        2.0     1.0  400.0       750.0  
3         SE  aracaju   67.0    2.0        1.0     1.0  395.0      1250.0  
4         SE  aracaju   70.0    3.0        1.0     1.0  380.0      1200.0  
5         SE  aracaju   65.0    3.0        1.0     1.0  375.0       800.0  
6         SE  aracaju   59.0    3.0        1.0     1.0  370.0       850.0  
...      ...      ...    ...    ...        ...     ...    ...         ...  
17028     GO  goiania   36.0    1.0        1.0     1.0  320.0      2950.0  
17064     GO  goiania   64.0    2.0        2.0     1.0  100.0      1200.0  
17099     GO  goiania   90.0    2.0        1.0     1.0  410.0       880.0  
17137     GO  goiania  173.0    2.0        3.0     2.0  480.0      1300.0  
17171     GO  goiania   43.0    1.0        1.0     1.0  340.0      2490.0  

[2659 rows x 10 columns]

In [30]:
# Salvando conjunto de dados no Google Drive

path = '/content/drive/MyDrive/data_science/imoveis/apto_capitais_data.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  df.to_csv(f)